src/domain/usecases/CopyProgramStageDataValuesUseCase.ts

```diff
@@ -0,0 +1,175 @@
+
+    async execute(options: CopyProgramStageDataValuesOptions): Promise<ProgramEvent[]> {
+        const { programStageId, dataElementIdPairs: idPairs, post, saveReport: reportPath } = options;
+
+        const rootOrgUnit = await this.orgUnitRepository.getRoot();
+        const dataElements = await this.dataElementsRepository.getByIds(idPairs.flat());
+        const dataElementPairs = this.mapDataElements(dataElements, idPairs);
+        const sourceIds = idPairs.map(([sourceId, _]) => sourceId);
+        const targetIds = idPairs.map(([_, targetId]) => targetId);
+
+        checkDataElementTypes(dataElementPairs);
+
+        const allEvents = await this.programEventsRepository.get({
+            programStagesIds: [programStageId],
+            orgUnitsIds: [rootOrgUnit.id],
+            orgUnitMode: "DESCENDANTS",
+        });
+
+        const applicableEvents = allEvents.filter(event =>
+            event.dataValues.some(dv => sourceIds.includes(dv.dataElement.id))
+        );
+
+        checkTargetDataValuesAreEmpty(applicableEvents, targetIds);
+
+        const eventsWithNewDataValues = this.copyEventDataValues(
+            applicableEvents,
+            sourceIds,
+            dataElementPairs
+        );
+
+        if (post) {
+            const result = await this.programEventsRepository.save(eventsWithNewDataValues);
+            if (result.type === "success") log.info(JSON.stringify(result, null, 4));
+            else log.error(JSON.stringify(result, null, 4));
+        } else {
+            const payload = { events: eventsWithNewDataValues };
+            const json = JSON.stringify(payload, null, 4);
+            const now = new Date().toISOString().slice(0, 19).replace(/:/g, "-");
+            const payloadPath = `copy-program-stage-data-values-${now}.json`;
+
+            fs.writeFileSync(payloadPath, json);
+            log.info(`Written payload (${eventsWithNewDataValues.length} events): ${payloadPath}`);
+        }
+
+        if (reportPath) {
+            this.saveReport(reportPath, dataElementPairs, programStageId, eventsWithNewDataValues);
+        }
+
+        return eventsWithNewDataValues;
+    }
+
+    private copyEventDataValues(
+        applicableEvents: ProgramEvent[],
+        sourceIds: string[],
+        dataElementPairs: DataElementPair[]
+    ) {
+        return applicableEvents.map(event => ({
+            ...event,
+            dataValues: event.dataValues.flatMap(dv => {
+                if (!sourceIds.includes(dv.dataElement.id)) return [dv];
+                const target = dataElementPairs.find(([source, _]) => source.id === dv.dataElement.id)?.[1];
+
+                if (!target)
+                    throw new Error(`Target data element not found for source id: ${dv.dataElement.id}`);
+
+                return [
+                    dv,
+                    {
+                        ...dv,
+                        dataElement: _.omit(target, "valueType"),
+                    },
+                ];
```

Comment:
The layout of conditional blocks is one of the most important things to be declarative. If you are able to put a single if/else if/else chain, it will look nicer. Proposal: 

```
                const target = dataElementPairs.find(([source, _]) => source.id === dv.dataElement.id)?.[1];

                if (!sourceIds.includes(dv.dataElement.id)) {
                    return [dv];
                } else if (!target) {
                    throw new Error(`Target data element not found for source id: ${dv.dataElement.id}`);
                } else {
                    return [dv, { ...dv, dataElement: _.omit(target, "valueType") }];
                }
```